File: iOS/KeioApp/KeioApp/presentation/Portal/PortalViewController.m
//
//  PortalViewController.m
//  KeioApp
//
//  Created by 荻原湧志 on 2018/05/30.
//  Copyright © 2018年 ogiwara. All rights reserved.
//

#import "PortalViewController.h"
#import "Constants.h"
#import "PortalRepository.h"
#import "PortalMessageDetailViewController.h"
#import "PortalMessageTableViewCell.h"
#import "AppDelegate.h"
#import "AccessKey.h"
#import "LoginDataRepository.h"

@interface PortalViewController ()
{
    NSMutableArray<PortalMessage *> *dataSource;
}



@end

@implementation PortalViewController
{
    id disposable;
}


- (void)viewDidLoad {
    [super viewDidLoad];
    dataSource = [[NSMutableArray alloc] init];
    
    [[LoginDataRepository shared] addObserver:self];
}

-(void)didLogin: (AccessKey *)accessKey{
    [[PortalRepository shared] index:accessKey compleion:^(Portal *portal) {
        [self->dataSource addObjectsFromArray: portal.messages];
        [self.tableView reloadData];
    }];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return dataSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    PortalMessageTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier: @"Cell"];
    
    [cell setData: dataSource[indexPath.row]];
    
    return cell;
}


- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    return 100;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender{
    if(![segue.identifier isEqualToString:@"segueShowDetail"]){
        return;
    }
    
    NSIndexPath *path = [self.tableView indexPathForSelectedRow];
    
    if(path == nil){
        return;
    }
    
    PortalMessageDetailViewController *ctrl = segue.destinationViewController;
    NSInteger index = [path row];
    ctrl.message = [dataSource objectAtIndex:index];
}

- (void)dealloc
{
 //   [disposable dispose];
}

@end

File: iOS Native/KeioClient/KeioClient/PortalViewController.swift
//
//  PortalViewController.swift
//  KeioClient
//
//  Created by 荻原湧志 on 2018/05/05.
//  Copyright © 2018年 ogiwara. All rights reserved.
//

import UIKit
import RxSwift

class PortalViewController: UITableViewController {

    var dataSource: [Message] = []
    var disposeable: Disposable? = nil
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        print("Portal VC has appear")
        
        //loadMessages()
    }
    
    /*
    private func loadMessages(){
        //guard let cookies = CookieRepository.shared.cookies else { return }
        
        CookieRepository.shared.subject
            .observeOn(MainScheduler.instance)
            //.timeout(<#T##dueTime: RxTimeInterval##RxTimeInterval#>, scheduler: <#T##SchedulerType#>)
        
        let repo = PortalRepository.init(cookies: cookies)
        repo.index
            .observeOn(MainScheduler.instance)
            .subscribe(
            onSuccess: { e in
                e.messages.forEach {
                    self.dataSource.append($0)
                }
                self.tableView.reloadData()
            },
            onError: { error in
                print(error)
            })
    }*/
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataSource.count
    }
    
    /*func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath as IndexPath) as! MessageTableViewCell
        cell.detail.text = dataSource[indexPath.row].title
        cell.title.text = dataSource[indexPath.row].detail
        return cell
    }*/
    
    deinit {
        disposeable?.dispose()
    }
}
